numpy/lib/shape_base.py

```diff
@@ -112,8 +112,9 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
         Ntot = product(outshape)
         holdshape = outshape
         outshape = list(arr.shape)
-        outshape[axis] = len(res)
+        outshape[axis] = len(res) if res.shape != () else 1
         outarr = zeros(outshape, asarray(res).dtype)
```

Comment:
Hmm, this line seems to be careful to not assume `res` is necessarily a numpy array or scalar (presumably it could be a list or so? I find this very odd...). If one wants to be super-sure, one could at the start of this non-scalar case write
```
res = asanyarray(res)
```
and then here the `asarray` could be dropped. That way one also ensures that `res` has a `shape`, `size`, and `__array_wrap__`.